internal/service/hub/handler/dsl/enforcer/score_maintainer.go

```diff
@@ -0,0 +1,154 @@
+package enforcer
+
+import (
+	"container/heap"
+	"sync"
+
+	"github.com/rss3-network/global-indexer/internal/service/hub/handler/dsl/model"
+	"github.com/samber/lo"
+)
+
+// priorityNodeQueue implements heap.Interface and holds NodeEndpointCaches.
+// It is used to maintain a priority queue of NodeEndpointCaches based on their scores and invalid counts.
+type priorityNodeQueue []*model.NodeEndpointCache
+
+func (pq priorityNodeQueue) Len() int {
+	return len(pq)
+}
+
+func (pq priorityNodeQueue) Less(i, j int) bool {
+	if pq[i].Score == pq[j].Score {
+		// If Scores are the same, return true if pq[i] has a smaller InvalidCount than pq[j]
+		return pq[i].InvalidCount < pq[j].InvalidCount
+	}
+	// Otherwise, return true if pq[i] has a greater score than pq[j]
+	return pq[i].Score > pq[j].Score
+}
+
+func (pq priorityNodeQueue) Swap(i, j int) {
+	pq[i], pq[j] = pq[j], pq[i]
+	pq[i].Index = i
+	pq[j].Index = j
+}
+
+func (pq *priorityNodeQueue) Push(x interface{}) {
+	n := len(*pq)
+	nodeEndpointCache := x.(*model.NodeEndpointCache)
+	nodeEndpointCache.Index = n
+	*pq = append(*pq, nodeEndpointCache)
+}
+
+func (pq *priorityNodeQueue) Pop() interface{} {
+	old := *pq
+	n := len(old)
+	nodeEndpointCache := old[n-1]
```

Comment:
did not handle zero len

Reply:
refer to https://pkg.go.dev/container/heap#example-package-PriorityQueue 
1. it's more proper to let it panic, rather than let the user get a nil return value.
2. we have already checked the length before use Pop.